File: C/layer.c
#include <malloc.h>
#include <time.h>

#include "loadingBar.h"
#include "layer.h"

double smoothstep(double w)
{
    if (w > 1)
    {
        w = 1;
    }
    else if (w < 0)
    {
        w = 0;
    }

    return w * w * (3.0 - 2.0 * w);
}



double interpolate(double a0, double a1, double w)
{
    return a0 + (a1 - a0) * smoothstep(w);
}





double dotGridGradient(int ix, int iy, double x, double y, gradientGrid* gradient_grid)
{
    double dx = x - (double) ix;
    double dy = y - (double) iy;

    vector* v = getVector(gradient_grid, ix, iy);

    return dx * v->x + dy * v->y;
}



double perlin(double x, double y, gradientGrid* gradient_grid)
{
    int x0 = (int) x;
    int y0 = (int) y;

    int x1 = x0 + 1;
    int y1 = y0 + 1;

    double sx = x - x0;
    double sy = y - y0;

    double n0 = dotGridGradient(x0, y0, x, y, gradient_grid);
    double n1 = dotGridGradient(x1, y0, x, y, gradient_grid);
    double ix0 = interpolate(n0, n1, sx);

    n0 = dotGridGradient(x0, y1, x, y, gradient_grid);
    n1 = dotGridGradient(x1, y1, x, y, gradient_grid);
    double ix1 = interpolate(n0, n1, sx);

    double value = interpolate(ix0, ix1, sy);

    return value;
}





double* getLayerValue(layer* layer, int width_idx, int height_idx)
{
    double* layer_value = NULL;

    if (layer != NULL)
    {
        int width = layer->width;
        int height = layer->height;

        if (width_idx < 0 || width_idx >= width)
        {
            printf("%sERROR : invalid width_idx = %d when reading layer value. Should be in range [0, %d]%s\n", RED_COLOR, width_idx, width - 1, DEFAULT_COLOR);
            return NULL;
        }
        
        if (height_idx < 0 || height_idx >= height)
        {
            printf("%sERROR : invalid height_idx = %d when reading layer value. Should be in range [0, %d]%s\n", RED_COLOR, height_idx, height - 1, DEFAULT_COLOR);
            return NULL;
        }

        layer_value = (layer->values) + height_idx * width + width_idx;
    }

    return layer_value;
}





layer* newLayerFromGradient(gradientGrid* gradient_grid, int size_factor, unsigned int display_loading)
{
    clock_t start_time = clock();

    int gradGridWidth = gradient_grid->width;
    int gradGridHeight = gradient_grid->height;

    // Size layer must be applied **between** gradient grids points.Thus, the layer dimensions are the following:
    int width = (gradGridWidth - 1) * size_factor;
    int height = (gradGridHeight - 1) * size_factor;

    // Initialization
    layer* new_layer = calloc(1, sizeof(layer));
    double* values = calloc(width * height, sizeof(double));

    new_layer->width = width;
    new_layer->height = height;
    new_layer->size_factor = size_factor;

    new_layer->gradient_grid = gradient_grid;

    new_layer->values = values;

    // Setting correct double values
    for (int i = 0; i < height; i++)
    {
        for (int j = 0; j < width; j++)
        {
            double* value = getLayerValue(new_layer, j, i);

            *value = perlin((double) j/size_factor, (double) i/size_factor, gradient_grid);

            if (display_loading != 0)
            {
                int nb_indents = display_loading - 1;

                char base_str[100] = "Generating layer...                ";

                predefined_loading_bar(j + i * width, width * height - 1, NUMBER_OF_SEGMENTS, base_str, nb_indents, start_time);
            }
        }
    }

    return new_layer;
}



layer* newLayer(int gradGrid_width, int gradGrid_height, int size_factor, unsigned int display_loading)
{
    int g_loading = display_loading;

    // if (display_loading != 0)
    // {
    //     indent_print(display_loading - 1, "Generating new layer...\n");
    //     g_loading += 1;
    // }

    // Size Factor should be set to match gradGrid_width - 1 and gradGrid_height - 1
    gradientGrid* gradient_grid = newRandomGradGrid(gradGrid_width, gradGrid_height, g_loading);

    return newLayerFromGradient(gradient_grid, size_factor, g_loading);
}





void writeLayerFile(layer* layer, char path[])
{
    FILE* f = NULL;

    f = fopen(path, "w");

    if (f != NULL)
    {
        fprintf(f, "Layer\n");

        // Writing the parameters
        int width = layer->width;
        int height = layer->height;
        int size_factor = layer->size_factor;

        fprintf(f, "width=%d\nheight=%d\nsize_factor=%d\n", width, height, size_factor);

        // Writing the values
        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                double value = *getLayerValue(layer, j, i);

                if (j != width - 1)
                {
                    fprintf(f, "% .8lf\t", value);
                }
                else
                {
                    fprintf(f, "% .8lf\n", value);
                }
            }
        }

        fclose(f);
    }
    else
    {
        printf("%sERROR : could not open file in writing mode at path '%s'%s\n", RED_COLOR, path, DEFAULT_COLOR);
        return;
    }
}



layer* readLayerFile(char path[200]);





void printLayer(layer* layer)
{
    int width = layer->width;
    int height = layer->height;

    printf("-------------------------------------------\n");
    printf("Printing layer of size = (%d, %d)\n\n", height, width);

    for (int i = 0; i < height; i++)
    {
        for (int j = 0; j < width; j++)
        {
            double* value = getLayerValue(layer, j, i);

            printf("%lf   ", *value);
        }
        printf("\n");
    }

    printf("-------------------------------------------\n");
}



void freeLayer(layer* layer)
{
    if (layer != NULL)
    {
        if (layer->values != NULL)
        {
            free(layer->values);
        }

        freeGradGrid(layer->gradient_grid);

        free(layer);
    }
}

File: Python/layer.py

#? To authorize type hints of class itself (ie Layer here)
#? answer found here : https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class
#from typing import Self                    #? Python 3.11+
from __future__ import annotations          #? Python 3.7+

import numpy as np
from gradientGrid import GradientGrid






class Layer:
    """
    Layer :
    -------------
    The layer class used to generate an altitude map based on a GradientGrid structure.
    """
    
    
    #? -------------------------- Static ------------------------- #
    
    
    ALT_PRINTING_DECIMALS = 4
    ALT_PRINTING_FORMAT = " {{: .{}f}} ".format(ALT_PRINTING_DECIMALS)
    
    
    @staticmethod
    def smoothstep(w: float) -> float:
        """A smoothstep function to smoothly go from 0 to 1.
        It uses the analytical function `3w² - 2w³`

        Args:
            `w` (float): the input value. Should be in `[0, 1]`

        Returns:
            float: the resulting smoothstep value.
        """
        
        if w <= 0:
            return 0.
        if w >= 1:
            return 1.
        
        return 3*w**2 - 2*w**3
    
    
    
    @staticmethod
    def interpolate(a0: float, a1: float, w: float) -> float:
        """A smooth interpolating function.

        Args:
            `a0` (float): the first value to interpolate from, which corresponds to `w = 0`.
            `a1` (float): the second value to interpolate from, which corresponds to `w = 1`.
            `w` (float): the smoothstep input. Should be between `0` and `1`.

        Returns:
            float: the interpolated value between `a0` and `a1`.
        """
        
        return a0 + (a1 - a0) * Layer.smoothstep(w)
    
    
    
    @staticmethod
    def dotGridGradient(ix: int, iy: int, x: float, y: float, grid: GradientGrid) -> float:
        """Computes the dot product between the (x - ix, y - iy) vector and the extracted vector from the given gradient grid.

        Args:
            `ix` (int): the width index to extract the vector from the gradient grid.
            `iy` (int): the height index to extract the vector from the gradient grid.
            `x` (float): the width position.
            `y` (float): the height position.
            `grid` (GradientGrid): the gradient grid to compute from.

        Returns:
            float: the computed dot product.
        """
        
        if int(ix) != ix or int(iy) != iy:
            raise ValueError("ix and iy must be integers! Are : {} and {}".format(ix, iy))
        
        ix = int(ix)
        iy = int(iy)
        
        dx = x - ix
        dy = y - iy

        v = grid.vectors[ix, iy]

        return dx * v[0] + dy * v[1]
    
    
    
    @staticmethod
    def perlin(x: float, y: float, grid: GradientGrid) -> float:
        """Computes the altitude value at position (x, y) from the given gradient grid.

        Args:
            `x` (float): the width position.
            `y` (float): the height position.
            `grid` (GradientGrid): the gradient grid to compute the altitude from.

        Returns:
            float: the final altitude value.
        """
        
        x0 = np.floor(x)
        x1 = x0 + 1
        y0 = np.floor(y)
        y1 = y0 + 1
        
        sx = x - x0
        sy = y - y0
        
        n0 = Layer.dotGridGradient(x0, y0, x, y, grid)
        n1 = Layer.dotGridGradient(x1, y0, x, y, grid)
        ix0 = Layer.interpolate(n0, n1, sx)
        n0 = Layer.dotGridGradient(x0, y1, x, y, grid)
        n1 = Layer.dotGridGradient(x1, y1, x, y, grid)
        ix1 = Layer.interpolate(n0, n1, sx)
        value = Layer.interpolate(ix0, ix1, sy)

        return value
    
    
    
    
    
    
    
    
    @staticmethod
    def generateLayerFromScratch(grid_width: int, grid_height: int, size_factor: int, adj_layers: tuple[Layer] = (None, None, None, None)) -> Layer:
        """Generates a new GradientGrid structure with the given parameters and then generate a new Layer structure from it.
        You can additionally specify adjacent layers during the generation to use their GradientGrid as neighbours for the new GradientGrid generation.

        Args:
            `grid_width` (int): the width to be used to generate an initial gradient grid.
            `grid_height` (int): the height to be used to generate an initial gradient grid.
            `size_factor` (int): the size factor to be used to generate the layer. The layer dimensions will be
                                 `(gradient_grid_dimensions - 1) * size_factor`.
            `adjacent_layers` (tuple[Layer], optional): the tuple of adjacent layers to generate a smooth transition in terrain with correct
                                                        boundary conditions.
                                                        The tuple is in order `(NORTH, EAST, SOUTH, WEST)`.
                                                        Set a layer to `None` if there are no neighbours in this direction.
                                                        Defaults to `(None, None, None, None)`.
        """
        
        if not (type(grid_width) is int and type(grid_height) is int and type(size_factor) is int and size_factor > 1):
            print("Invalid parameters : grid dimensions and size_factor should be integers. Moreover, size_factor should be > 1.")
            return
        
        adj_list = [None, None, None, None]
        for i in range(len(adj_layers)):
            layer = adj_layers[i]
            if type(layer) is Layer:
                adj_list[i] = layer.grid
        adj_tuple = tuple(adj_list)
        
        
        grid: GradientGrid = GradientGrid(grid_width, grid_height, adjacent_grids=adj_tuple)
        
        return Layer(grid=grid, size_factor=size_factor)
    
    
    
    @staticmethod
    def write(path: str, layer: Layer) -> None:
        #TODO
        pass
    
    
    
    @staticmethod
    def read(path: str) -> Layer:
        #TODO
        return None
    
    
    #? ------------------------ Instances ------------------------ #
    
    
    def __init__(self, grid: GradientGrid, size_factor: int) -> None:
        """Initializes a new layer instance from an already existing GradientGrid and a size factor.
        If no gradient grid are already generated, please use the designated static method `generateLayerFromScratch` to generate the
        appropriate GradientGrid and then generate the layer structure.

        Args:
            `grid` (GradientGrid): the gradient grid to build the layer from.
            `size_factor` (int): the size factor to be used to generate the layer. The layer dimensions will be
                                 `(gradient_grid_dimensions - 1) * size_factor`.
        """
        
        self.grid = None
        self.width = None
        self.height = None
        
        self.size_factor = None
        
        self.altitude = None
        
        
        if not (type(grid) is GradientGrid and type(size_factor) is int and size_factor > 1):
            print("Invalid parameters. Either gradient grid is None, size_factor is not an integer, or is <= 1.")
            return
        
        
        self.grid = grid
        self.size_factor = size_factor
        
        self.regenerate()
    
    
    
    def __str__(self) -> str:
        """Returns the string representation of the Layer.
        Used by the print function when called with a Layer.

        Returns:
            `str`: the string representation of the Layer.
        """
        final_str = "-----------------------------------------\n"
        final_str += "Layer of dimensions {} x {} :\n".format(self.width, self.height)
        for i in range(self.height):
            
            for j in range(self.width):
                alt = self.altitude[i, j]
                final_str += Layer.ALT_PRINTING_FORMAT.format(alt)
            
            final_str += "\n"
        final_str += "-----------------------------------------\n"
        
        return final_str
    
    
    
    def regenerate(self) -> None:
        """Regenerates the altitude values of the layer based on its `grid` and `size_factor` parameters.
        """
        
        if not (type(self.grid) is GradientGrid and type(self.size_factor) is int and self.size_factor > 1):
            print("Invalid parameters. Either gradient grid is None, size_factor is not an integer, or is <= 1.")
            return
        
        self.width = (self.grid.width - 1) * self.size_factor
        self.height = (self.grid.height - 1) * self.size_factor
        
        self.altitude = np.zeros((self.height, self.width))
        
        for i in range(self.height):
            for j in range(self.width):
                
                self.altitude[i, j] = Layer.perlin(i/self.size_factor, j/self.size_factor, self.grid)





# --------------------------------- # Testing Layer # --------------------------------- #

if __name__ == "__main__":
    print("# --------------------------------- # Testing Layer # --------------------------------- #\n")
    
    GradientGrid.setRandomSeed("seed")
    
    factor = 4
    grid = GradientGrid(5, 5)
    layer = Layer(grid=grid, size_factor=factor)
    print(layer)
    
    
    
    print("\nTrying to regenerate : ")
    grid.regenerate()
    layer.regenerate()
    print(layer)
    
    
    
    print("Trying to generate a layer at the north of the first one :")
    layern = Layer.generateLayerFromScratch(grid_width=5, grid_height=5, size_factor=factor, adj_layers=(None, None, layer, None))
    print(layer.grid)
    print(layern.grid)
    print(layern)
